### ISSUE TITLE
interaction between tooltips and gridplots

### ISSUE THREAD

**Issue by troianow** [bokeh/bokeh#4930]
I tried to build a grid plot involving tooltips. The result was that, if the tooltips display was not contained in the figure associated to it, it was truncated by the figure to the right and below as in the following example. However, it does not get truncated by figures above or to the left:
## Screenshot

This one is truncated
![image](https://cloud.githubusercontent.com/assets/1373343/17448385/fe7767ae-5b21-11e6-9648-ab558a80e623.png)

This one is not:
![image](https://cloud.githubusercontent.com/assets/1373343/17448434/296f805e-5b22-11e6-8358-45affb0f6177.png)
## Code

``` python
import numpy
from bokeh.plotting import figure, show, output_file
from bokeh.plotting import figure, output_file, show, ColumnDataSource
from bokeh.sampledata.iris import flowers
from bokeh.io import gridplot
from bokeh.models import HoverTool

df = flowers.copy()
df['cat_col'] = numpy.round(flowers.petal_width * 2)

colormap = {'setosa': 'red', 'versicolor': 'green', 'virginica': 'blue'}
colors = [colormap[x] for x in df['species']]

result = []
for species in df['species'].unique():
    sub_res = []
    for cat in df['cat_col'].unique():
        sub_df = df[(df['species'] == species) & (df['cat_col'] == cat)]
        colors = [colormap[x] for x in sub_df['species']]
        hoverdata = ['species:{} - cat:{}'.format(species, cat) for x in sub_df['species']]
        source = ColumnDataSource(data=dict(hovercol=hoverdata))

        hover = HoverTool(tooltips=[("Hover col", "@hovercol"),])
        p = figure(tools=[hover])
        p.circle(sub_df["petal_length"], sub_df["petal_width"],
                 color=colors, fill_alpha=0.2, size=10, source=source)
        sub_res.append(p)
    result.append(sub_res)


output_file("iris.html", title="iris.py example")

show(gridplot(result, plot_width=200, plot_height=200))
```
## Bokeh version

Python version      :  3.4.3 |Continuum Analytics, Inc.| (default, Dec  1 2015, 11:59:33) [MSC v.1600 32 bit (Intel)]
IPython version     :  4.0.0
Bokeh version       :  0.12.1

**Comment by birdsarah** [comment]
Hi @troianow - there is an open issue for that here: https://github.com/bokeh/bokeh/issues/2094. 
However this issue much better documents the problem, so I'm going to close that one in favor of this one.

**Comment by birdsarah** [comment]
Now the bad news. If I'm recalling correctly, this is a hard problem to fix because of the way that z-index nesting works - https://css-tricks.com/almanac/properties/z/z-index/

**Comment by troianow** [comment]
Hi @birdsarah - Is the tooltip from jquery-ui? Looking at the css, it seems like the .bk-ui-tooltip defines an absolute z-index of 9999. Yet, it looks like the order of creation of the figures creates a hierarchy in the z-index. Does this mean the tooltip's z-index is relative to the already existing figures? Could it be defined at the level of the grid, after the figures are initialized? Am I missing something?

**Comment by mattpap** [comment]
>  Is the tooltip from jquery-ui?

jquery-ui isn't used by core bokehjs.

> Does this mean the tooltip's z-index is relative to the already existing figures? Could it be defined at the level of the grid, after the figures are initialized?

You can observe this as tooltips aligned to the left (top) appear normally and those aligned to the right (bottom) don't. As @birdsarah mentioned, `z-index` is pretty tricky to deal with, so adding tooltips to a common parent element (not even a grid, but bokeh document's element) may a valid solution. Currently we add tooltips per cavnas wrapper element, because this is just a natural solution, which is supposed to work for one or multiple plots without extra logic (assuming `z-index` is absolute, which is not true).

**Comment by bryevdv** [comment]
Right the 9999 `z-index` was my previous attempt to fix this problem. I was not aware of the nested `z-index` CSS rules. Give that I think the only solution is as described, to put the tooltip on something that is a parent to everything. I just don't think it will be trivial to do, unfortunately.

**Comment by birdsarah** [comment]
Another option which doesn't fix it, but may help, would be to implement hover in a fixed position, like on this leaflet demo: http://leafletjs.com/examples/choropleth.html

<img width="763" alt="screen shot 2016-08-08 at 2 43 59 pm" src="https://cloud.githubusercontent.com/assets/1796208/17497402/91afd370-5d76-11e6-831b-ac910b3aaff1.png">

**Comment by bryevdv** [comment]
@mattpap @birdsarah I think this another "must-fix for 1.0" candidate

**Comment by birdsarah** [comment]
@bryevdv I would put it in the "must have a good workaround". Any of the thoughts I've had to fix seem insanely hard to me. I'm not sure what to do about this.

**Comment by mmcguffi** [comment]
I see as mentioned in #1744 that the hovertool now wraps around on the right, but when the hover tool "points" to the left, it still overflows.

eg:
<img width="549" alt="Screen Shot 2020-02-23 at 2 46 08 PM" src="https://user-images.githubusercontent.com/17573404/75119770-85782180-564b-11ea-8619-1f04153f1ce4.png">
<img width="689" alt="Screen Shot 2020-02-23 at 2 46 30 PM" src="https://user-images.githubusercontent.com/17573404/75119773-88731200-564b-11ea-8a22-e59cb6176bd4.png">

Are there any fixes for this?

**Comment by bryevdv** [comment]
@mmcguffi possibly https://github.com/bokeh/bokeh/pull/9634 but it is not in a release yet, so you would need to test with a local build from master, or provide a test case we can actually run.

**Comment by mmcguffi** [comment]
@bryevdv ah yes, this appears to be exactly what is happening to me. I probably wont publicly release what I am developing for at least an other couple months, so I will probably just wait for the 2.0 release. Thanks!

**Comment by bryevdv** [comment]
@mmcguffi just trying to understand clearly, as you saying you tried things out yourself and they were fixed? FYI I put an RC up on pypi for easier testing 

https://pypi.org/project/bokeh/2.0.0rc1/

Also one the `bokeh/dev` channel on anaconda.org

**Comment by mmcguffi** [comment]
@bryevdv I did not actually test it empirically -- the documented behavior just seems to completely track with my issues. Thank you for solving my laziness and throwing it up on PyPI though -- I just tried the v2.0rc1 and it does indeed fix the problem.  

<img width="495" alt="Screen Shot 2020-02-23 at 7 55 41 PM" src="https://user-images.githubusercontent.com/17573404/75124692-8bcfc300-5676-11ea-9f3d-dc1866b78016.png">

However, this RC version actually completely kills my visualization when run through [Streamlit](https://github.com/streamlit/streamlit). No rendering whatsoever and does not throw any errors. Not sure what is happening here and I not sure who's end (Bokeh/Streamlit) the problem would be on.

**Comment by bryevdv** [comment]
@mmcguffi I am afraid I don't know anything about the details of streamlit. This is a major number bump, and there are breaking changes. So it's not completely unexpected that downstream projects might need adjustments. cc @MarcSkovMadsen

**Comment by andfanilo** [comment]
Hello,

I'm quickly chiming in from the Streamlit side where @mmcguffi [initiated the conversation](https://discuss.streamlit.io/t/bokeh-2-0-potentially-broken-in-streamlit/2025) about Bokeh 2.0.0rc1 not working in Streamlit, hopefully I don't sidetrack your initial conversation too much...

Currently, on the Python side Streamlit emits the bokeh chart into a JSON with `bokeh.embed.json_item` and then on the JS side uses this JSON in `bokehjs.embed.embed_item`.

My guess is, if Bokeh on the Python side is upgraded to 2.0.0, then BokehJS on the JS side should also be 2.0.0 to be able to parse the JSON but I could not find a corresponding version for BokehJS to test, and its only a guess...also the embed_item from BokehJS did not throw any error on reading the upgraded JSON item and that felt odd.

* Do you plan to also release a BokehJS 2.0.0rc1 on npm for downstream projects to test Bokeh embedding ? 
* Do you eventually have a migration guide to 2.0.0 ? 

Don't hesitate to point me to the correct issue or discussion for this !

**Comment by philippjfr** [comment]
@andfanilo I'll let @bryevdv field most of those questions but there have indeed been changes to the way `json_item` works so I'm not surprised non-matching Python and JS versions won't work together.

**Comment by bryevdv** [comment]
> My guess is, if Bokeh on the Python side is upgraded to 2.0.0, then BokehJS on the JS side should also be 2.0.0 to be able to parse the JSON but I could not find a corresponding version for BokehJS to test

Yes versions definitely JS/Python need to match, in general. Not just between major releases, either. Are you intending to hard code a fixed BokehJS version but accept JSON output from arbitrary Python Bokeh versions? I would not regard that as reliable. 

> Do you plan to also release a BokehJS 2.0.0rc1 on npm for downstream projects to test Bokeh embedding ? 

I had not planned on it. Until now, we only publish full releases to NPM. This is certainly something we can consider changing in the future, especially if there is increased interest in the NPM package. Given the compressed schedule to 2.0 this week I am not sure it's worth it for me to scramble to do something different in the middle of release week. But we can have a very quick 2.0.1 if there are any follow-on issues. 

> Do you eventually have a migration guide to 2.0.0 ?

Each release has its own (if any) migration notes:

https://docs.bokeh.org/en/dev/docs/releases.html

That said, while we do our best, we only put in what we think to put in based on the use cases we know about, so there may be things that are missed for some people.

**Comment by andfanilo** [comment]
Thank you both of you @bryevdv @philippjfr I transferred the info back to the Streamlit devs so we can look at what to do next. 

Best of luck for the upcoming release !

**Comment by bryevdv** [comment]
This was fixed at some point (result with Bokeh 3.0)

<img width="462" alt="Screenshot 2022-11-02 at 19 34 50" src="https://user-images.githubusercontent.com/1078448/199637586-eebd454e-4242-4100-a1bd-8ca241436636.png">

Updated MRE:
```
import numpy
from bokeh.plotting import figure, show, ColumnDataSource
from bokeh.sampledata.iris import flowers
from bokeh.layouts import gridplot
from bokeh.models import HoverTool

df = flowers.copy()
df['cat_col'] = numpy.round(flowers.petal_width * 2)

colormap = {'setosa': 'red', 'versicolor': 'green', 'virginica': 'blue'}
colors = [colormap[x] for x in df['species']]

result = []
for species in df['species'].unique():
    sub_res = []
    for cat in df['cat_col'].unique():
        sub_df = df[(df['species'] == species) & (df['cat_col'] == cat)]
        colors = [colormap[x] for x in sub_df['species']]
        hoverdata = ['species:{} - cat:{}'.format(species, cat) for x in sub_df['species']]
        source = ColumnDataSource(data=sub_df)
        source.data["hovercol"] = hoverdata
        source.data["colors"] = colors

        hover = HoverTool(tooltips=[("Hover col", "@hovercol"),])
        p = figure(tools=[hover])
        p.circle("petal_length", "petal_width",
                 color="colors", fill_alpha=0.2, size=10, source=source)
        sub_res.append(p)
    result.append(sub_res)

show(gridplot(result, width=200, height=200))
```

**Comment by mattpap** [comment]
Was a regression test committed for this issue? If not, then this is not complete and likely will be broken again.
